src/main/java/com/github/manolo8/darkbot/gui/drawables/ZonesDrawer.java

```diff
@@ -0,0 +1,134 @@
+
+    private final ConfigSetting<Integer> zoneResolution;
+    private final ConfigSetting<Boolean> roamingSequential;
+
+
+    public ZonesDrawer(EntitiesAPI entities, ConfigAPI config, StarSystemAPI starSystem) {
+        this.config = config;
+        this.starSystem = starSystem;
+
+        this.mists = entities.getMists();
+        this.barriers = entities.getBarriers();
+
+        this.zoneResolution = config.requireConfig("bot_settings.other.zone_resolution");
+        this.roamingSequential = config.requireConfig("general.roaming.sequential");
+    }
+
+    @Override
+    public void onDraw(MapGraphics mg) {
+        drawZones(mg);
+        drawCustomZones(mg);
+    }
+
+    public void drawZones(MapGraphics mg) {
+        for (Barrier barrier : barriers) {
+            if (!barrier.use()) return;
+
+            mg.setColor("barrier");
+            mg.drawArea(barrier.getZoneArea(), true);
+
+            mg.setColor("barrier_border");
+            mg.drawArea(barrier.getZoneArea(), false);
+        }
+
+        mg.setColor("no_cloack");
+        for (Mist mist : mists) {
+            mg.drawArea(mist.getZoneArea(), true);
+        }
+    }
+
+    public void drawCustomZones(MapGraphics mg) {
+        if (!mg.hasDisplayFlag(DisplayFlag.ZONES)) return;
+
+        mg.setColor("prefer");
+        drawCustomZone(mg, config.getLegacy().getPreferredZone(starSystem.getCurrentMap()));
+
+        if (roamingSequential.getValue())
+            drawCustomZonePath(mg, config.getLegacy().getPreferredZone(starSystem.getCurrentMap()));
+
+        mg.setColor("avoid");
+        drawCustomZone(mg, config.getLegacy().getAvoidedZone(starSystem.getCurrentMap()));
+
+        mg.setColor("safety");
+        for (SafetyInfo safety : config.getLegacy().getSafeties(starSystem.getCurrentMap())) {
+            if (safety.getRunMode() == SafetyInfo.RunMode.NEVER
+                || !safety.getEntity().filter(Entity::isValid).isPresent()) continue;
+
+            drawSafeZone(mg, safety);
+        }
+    }
+
+    public void drawCustomZone(MapGraphics mg, ZoneInfo zoneInfo) {
+        if (zoneInfo == null) return;
+        for (int x = 0; x < zoneInfo.getResolution(); x++) {
+            for (int y = 0; y < zoneInfo.getResolution(); y++) {
+                if (!zoneInfo.get(x, y)) continue;
+
+                Point pos = Point.of(gridToMapX(mg, x), gridToMapY(mg, y));
+                mg.drawRect(pos, gridToMapX(mg, x + 1) - pos.x(), gridToMapY(mg, y + 1) - pos.y(), true);
+            }
+        }
+    }
+
+    public void drawCustomZonePath(MapGraphics mg, ZoneInfo zoneInfo) {
+        if (!(zoneInfo instanceof com.github.manolo8.darkbot.config.ZoneInfo)) return;
+
+        List<com.github.manolo8.darkbot.config.ZoneInfo.Zone> sortedZones =
+                ((com.github.manolo8.darkbot.config.ZoneInfo) zoneInfo).getSortedZones();
+
+        Area.Rectangle mapBounds = starSystem.getCurrentMapBounds();
+
+        //convert it to actual list of locatables?
```

Comment:
probably best to do that, yeah
just add them all to a list and draw it as a poly